contracts/PangolinExchange.sol

```diff
@@ -0,0 +1,166 @@
+import "./IAssetsExchange.sol";
+import "./SupportedAssets.sol";
+
+/**
+ * @title PangolinExchange
+ * @dev Contract allows user to invest into an ERC20 token
+ * This implementation uses the Pangolin DEX
+ */
+contract PangolinExchange is Ownable, IAssetsExchange {
+  /* ========= STATE VARIABLES ========= */
+  IPangolinRouter pangolinRouter;
+  SupportedAssets supportedAssets;
+
+  /* ========= CONSTRUCTOR ========= */
+
+  constructor (address _pangolinRouter, SupportedAssets _supportedAssets) {
+    pangolinRouter = IPangolinRouter(_pangolinRouter);
+    supportedAssets = _supportedAssets;
+  }
+
+  /* ========= MODIFIERS ========= */
+
+  modifier RefundRemainder {
+    _;
+    if (address(this).balance > 0) {
+      (bool success,) = msg.sender.call{value : address(this).balance}("");
+      require(success, "Refund failed");
+    }
+  }
+
+
+  function transferBack(bytes32 _asset) external override {
+    address tokenAddress = supportedAssets.getAssetAddress(_asset);
+    IERC20 token = IERC20(tokenAddress);
+    token.transfer(msg.sender, token.balanceOf(address(this)));
+  }
+
+
+  /**
+   * Buys selected ERC20 token with AVAX using the Pangolin DEX
+   * Refunds unused AVAX to the msg.sender
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountOut amount of the ERC20 token to be bought
+  **/
+  function buyAsset(bytes32 _token, uint256 _exactERC20AmountOut) payable external override RefundRemainder {
+    require(_exactERC20AmountOut > 0, "Amount of tokens to buy has to be greater than 0");
+
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    uint256 amountIn = getEstimatedAVAXForERC20Token(_exactERC20AmountOut, tokenAddress);
+
+    require(msg.value >= amountIn, "Not enough funds provided");
+
+    pangolinRouter.swapAVAXForExactTokens{value : msg.value}(_exactERC20AmountOut, getPathForAVAXtoToken(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenPurchase(msg.sender, _exactERC20AmountOut, block.timestamp);
+  }
+
+
+  /**
+   * Sells selected ERC20 token for AVAX
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountIn amount of the ERC20 token to be sold
+   * @dev _minAvaxAmountOut minimum amount of the AVAX token to be bought
+  **/
+  function sellAsset(bytes32 _token, uint256 _exactERC20AmountIn, uint256 _minAvaxAmountOut) external override RefundRemainder {
+    require(_exactERC20AmountIn > 0, "Amount of tokens to sell has to be greater than 0");
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    IERC20 token = IERC20(tokenAddress);
+    token.approve(address(pangolinRouter), _exactERC20AmountIn);
+    pangolinRouter.swapExactTokensForAVAX(_exactERC20AmountIn, _minAvaxAmountOut, getPathForTokenToAVAX(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenSell(msg.sender, _exactERC20AmountIn, block.timestamp);
+  }
+
+  /* ========== RECEIVE AVAX FUNCTION ========== */
+  receive() external payable {  }
```

Comment:
This can only happen in an event when someone directly sends funds to this contract or when we sell an asset and receive AVAX from the PangolinRouter. The first case is rather rare and the second one is already covered with the `TokenSell` event so I guess there is no real need to add this one as it would duplicate the `TokenSell` events.